### ISSUE TITLE
Need Arduino Mega examples / help?

### ISSUE THREAD

**Issue by scriptguru** [bookmanarchive/rom-dumper#1]
I've previously successfully dumped a parallel ROM (obscure model, unsupported by any programmers I had access to) using Arduino Mega.
Even though I've used `digitalRead` / `digitalWrite` (which are very slow compared to more efficient methods), it dumped 2MB file to an SD card in very reasonable time (well under a minute, if I recall correctly).
Mega has plenty of digital i/o pins, is easy to use, and easy to interface with a primitive SD card reader/writer.
The source is not shared anywhere because it didn't look valuable enough to share, but if it helps your project I could share it.
There is not much code, in fact.

Additionally I'd be interested in possibility to create a rewritable cart. I have DBS-2, don't have a Bookman device, but it looks like the technology is similar.

**Comment by jsyang** [comment]
Yes, I think the Arduino code should be fairly straightforward. No particular requirement for the Mega specifically, it's just the easiest thing I have on hand. It's more the HW setup for it; was thinking of it like this:
- Arduino or other 3v3 (all the cards run on 3V3, so will need a level shifter otherwise) SBC / MC with at least 12 GPIOs 
  - 2 for the shift registers
  - 8 for the data bus
  - some spares left over
- 4 x 8-bit shift registers to set the OE / CE / Addr pins
- 8 x GPIOs to read the data from the ROM after Addr has been set
- Dump to serial port directly

That would ensure minimal requirements for anybody to replicate the setup. The hardest part would be cannibalizing a reader's card slot for the interface.

Re: the DBS-2, I also agree that the software would be extremely similar. I think both DBS (1+2) and Bookman1 shared much of the same tech (6502) and possibly even the same-ish layout. I have pulled one of the DBS cards apart, seems the same on the inside: up to 2 COB (chip-on-board, "glob-top") mask ROMs hooked up to the card pins. Though haven't figured out the pinout in the same way that I have for Bookman units.

Am happy to accept any help here, thanks for reaching out!

**Comment by scriptguru** [comment]
With Mega I don't see why shift registers would be needed. An SD card reader/writer module is like $1 on AliExpress, so I wouldn't consider it a complex or expensive setup. Literally, just Mega, SD card module, and the card itself (which also can be found almost for free, or literally for free). Much more convenient than reading from a serial port IMO.

Yes, unfortunately the pinout is different on DBS. I was about to try reading a card, but compared the pinout first to the one provided on your website, and it wasn't matching. I'll see if I can find it out using the same method as you did.

**Comment by jsyang** [comment]
Sorry, your idea is sound. 👍 It didn't register immediately that with an SD slot it could dump standalone.

For the pinouts, be sure to check the operating voltages, I suspect some small minority of cards may be 5V, rather than the usual 3V3. I've been unable to dump a handful of them (everything else checked) and suspect this was the reason.

Don't know if you've read it but I have a roadmap outlined in the repo as well. Feel free to add to this for DBS
